Bugfix/zimbra 128 contact freq graph user time zone fix

varunrisbud opened the pull request.

varunrisbud committed:
Calculate contact frequency graph start time as per users current timezone

Timezone and offset are used interchangeably in this commit message.

As per the discussion with the product and front-end team the
contact frequency graph start time should be calculated as per user's current timezone.

User's current timezone will be provided in the soap request in form of offset in minutes.

If the user's current timezone is not provided then user configured timezone will be used.

User's configured timezone is determined using
com.zimbra.cs.mailbox.calendar.Util::getAccountTimeZone(account).

varunrisbud committed:
Added offsetInMinutes as parameter to GetContactFrequencyRequest.
Changed the range start time from UTC to milliseconds from epoch in response.

Users current timezone will be passed as offsetInMinutes in the to GetContactFrequencyRequest.

Changed the timestamps in the response from UTC to milliseconds from epoch timestamps.

Modified ZMailboxUtil to accept graph frequencies as separated by space instead of comma.

varunrisbud committed:
Changed GetContactFrequencyRequest time range attribute to be passed as values concatenated together.

In order to keep consistency with the API standards values for time range
can be passed as any combination of "dwm" like "dw", "wm", "d", etc.

In the ZMailboxUtil for demo purposes changed the format of timestamps
in the graph data from Epoch in milliseconds to UTC timestamps.(ISO-8601 format).

varunrisbud committed:
Minor fix to the testcase

ilyusha commented on soap/src/java/com/zimbra/soap/mail/message/GetContactFrequencyRequest.java:
```diff
@@ -21,16 +26,25 @@ public GetContactFrequencyRequest(String email, String frequencyBy) {
     private String frequencyBy;
 
+    /**
+     * @zm-pai-field-description offset in minutes of user's current timezone.
```
typo: pai -> api

ilyusha commented on store/src/java/com/zimbra/cs/event/SolrEventStore.java:
```diff
@@ -174,39 +180,49 @@ private String getAggregationBucketForContactFrequencyGraphTimeRange(ContactAnal
+        int minutes = Math.abs(offsetInMinutes) % 60;
+        String timeFormat = "%02d";
+        StringBuilder solrTimeZone = new StringBuilder();
+        solrTimeZone.append("GMT").append(sign).append(String.format(timeFormat, hours)).append(String.format(timeFormat, minutes));
```
This string has no conditionals, so you could build this string with String.format() without mixing in StringBuilder calls: `String.format("GMT%s%02d%02d", sign, hours, minutes)`

ilyusha commented on store/src/java/com/zimbra/cs/zclient/ZMailboxUtil.java:
```diff
@@ -2677,14 +2678,26 @@ private void doGetContacts(String[] args) throws ServiceException {
+            Instant utcTimestamp = Instant.ofEpochMilli(Long.parseLong(unixTimestamp));
+            dataPoint.setLabel(utcTimestamp.toString());
+        }
+        stdout.println("Epoch milliseconds are included in the actual response. UTC timestamp shown here are just for demo");
```
I'd shorten this to something like "Date labels shown in UTC; other clients may show the browser's timezone instead"

varunrisbud committed:
Made small changes as per code review

Code review - https://github.com/varunrisbud/zm-mailbox/pull/8

GetContactFrequencyRequest - fixed a typo.

SolrEventStore - Replaced String Builder in getSolrTimeZone() with
String.Format().

ZMailboxUtil - Changed a print statement.

varunrisbud committed:
Fixed stylictic issue raised by codacy-bot
